File: apps/mobile-app/lib/features/students/services/students_service.dart
/// ============================================================================
/// STUDENTS SERVICE — Academia Hub Mobile
/// ============================================================================
///
/// Service CRUD pour le module Élèves, miroir du web app (students.service.ts).
///
/// Étend [BaseCrudService] et fournit les méthodes spécifiques au module :
/// - CRUD de base (getAll, getById, create, update, delete)
/// - Admissions (CRUD + submit, decide, convert)
/// - Inscriptions / Enrollments
/// - Pré-inscription
/// - Dossier & Historique
/// - Statistiques & Orion (KPIs, Alertes)
/// - Photos & Gardiens
/// - Identifiants (Matricules, Cartes d'identité)
/// - Recherche par matricule
///
/// Pattern offline-first :
/// - Lectures → apiGetWithFallback (API avec fallback local)
/// - Écritures → apiPostWithOfflineFallback / apiUpdateWithOfflineFallback
///   (API avec outbox pattern hors ligne)
/// ============================================================================

import '../../../core/crud/base_crud_service.dart';
import '../../../core/network/api_config.dart';
import '../../../core/network/api_result.dart';
import '../../../core/offline/offline_service.dart';

class StudentsService extends BaseCrudService {
  // ─── Singleton ────────────────────────────────────────────────────────────

  static StudentsService? _instance;

  /// Instance singleton du service.
  static StudentsService get instance => _instance ??= StudentsService._();

  StudentsService._() : super(
    endpoint: ApiConfig.students,
    collection: 'students',
    entityType: SyncEntityType.student,
  );

  // ─── CRUD de base ────────────────────────────────────────────────────────

  /// Récupère tous les élèves avec filtres optionnels.
  ///
  /// Miroir de `studentsService.getAll()` du web app.
  Future<ApiResult<List<Map<String, dynamic>>>> getAll({
    Map<String, dynamic>? params,
  }) => super.getAll(params: params);

  /// Récupère un élève par son identifiant.
  ///
  /// Miroir de `studentsService.getById(id)` du web app.
  Future<ApiResult<Map<String, dynamic>>> getById(String id) =>
      super.getById(id);

  /// Crée un nouvel élève.
  ///
  /// Miroir de `studentsService.create(data)` du web app.
  /// Hors ligne → outbox pattern (SyncEntityType.student).
  Future<ApiResult<Map<String, dynamic>>> create(
    Map<String, dynamic> data,
  ) => super.create(data);

  /// Met à jour un élève existant.
  ///
  /// Miroir de `studentsService.update(id, data)` du web app.
  /// Hors ligne → outbox pattern.
  Future<ApiResult<Map<String, dynamic>>> update(
    String id,
    Map<String, dynamic> data,
  ) => super.update(id, data);

  /// Supprime un élève.
  ///
  /// Miroir de `studentsService.delete(id)` du web app.
  /// Hors ligne → soft delete + outbox pattern.
  Future<ApiResult<void>> delete(String id) => super.delete(id);

  // ─── Admissions ──────────────────────────────────────────────────────────

  /// Récupère toutes les admissions.
  ///
  /// Miroir de `studentsService.getAdmissions()` du web app.
  Future<ApiResult<List<Map<String, dynamic>>>> getAdmissions({
    Map<String, dynamic>? params,
  }) =>
      apiGetWithFallback<List<Map<String, dynamic>>>(
        ApiConfig.studentAdmissions,
        params: params,
        localCollection: 'students',
        localFilters: {'type': 'admission'},
      );

  /// Récupère une admission par son identifiant.
  ///
  /// Miroir de `studentsService.getAdmissionById(id)` du web app.
  Future<ApiResult<Map<String, dynamic>>> getAdmissionById(String id) =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentAdmissionById(id),
        localCollection: 'students',
        localFilters: {'type': 'admission', 'id': id},
      );

  /// Crée une nouvelle admission.
  ///
  /// Miroir de `studentsService.createAdmission(data)` du web app.
  /// Hors ligne → outbox pattern.
  Future<ApiResult<Map<String, dynamic>>> createAdmission(
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentAdmissions,
        data,
      );

  /// Met à jour une admission existante.
  ///
  /// Miroir de `studentsService.updateAdmission(id, data)` du web app.
  /// Hors ligne → outbox pattern.
  Future<ApiResult<Map<String, dynamic>>> updateAdmission(
    String id,
    Map<String, dynamic> data,
  ) =>
      apiUpdateWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentAdmissionById(id),
        id,
        data,
      );

  /// Soumet une admission pour validation.
  ///
  /// Miroir de `studentsService.submitAdmission(id)` du web app.
  Future<ApiResult<Map<String, dynamic>>> submitAdmission(String id) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentAdmissionSubmit(id),
        {'id': id},
      );

  /// Prend une décision sur une admission (accepter / refuser).
  ///
  /// Miroir de `studentsService.decideAdmission(id, data)` du web app.
  Future<ApiResult<Map<String, dynamic>>> decideAdmission(
    String id,
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentAdmissionDecide(id),
        data,
      );

  /// Convertit une admission en inscription définitive.
  ///
  /// Miroir de `studentsService.convertAdmission(id)` du web app.
  Future<ApiResult<Map<String, dynamic>>> convertAdmission(String id) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentAdmissionConvert(id),
        {'id': id},
      );

  // ─── Enrollments / Inscriptions ──────────────────────────────────────────

  /// Récupère toutes les inscriptions.
  ///
  /// Miroir de `studentsService.getEnrollments()` du web app.
  Future<ApiResult<List<Map<String, dynamic>>>> getEnrollments({
    Map<String, dynamic>? params,
  }) =>
      apiGetWithFallback<List<Map<String, dynamic>>>(
        ApiConfig.studentEnrollments,
        params: params,
        localCollection: 'students',
        localFilters: {'type': 'enrollment'},
      );

  /// Crée une nouvelle inscription.
  ///
  /// Miroir de `studentsService.createEnrollment(data)` du web app.
  Future<ApiResult<Map<String, dynamic>>> createEnrollment(
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentEnrollments,
        data,
      );

  /// Pré-inscrit un élève.
  ///
  /// Miroir de `studentsService.preRegister(data)` du web app.
  Future<ApiResult<Map<String, dynamic>>> preRegister(
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentPreRegister,
        data,
      );

  // ─── Dossier & Historique ────────────────────────────────────────────────

  /// Récupère le dossier complet d'un élève.
  ///
  /// Miroir de `studentsService.getDossier(id)` du web app.
  Future<ApiResult<Map<String, dynamic>>> getDossier(String id) =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentDossier(id),
        localCollection: 'students',
        localFilters: {'id': id},
      );

  /// Récupère l'historique d'un élève.
  ///
  /// Miroir de `studentsService.getHistory(id)` du web app.
  Future<ApiResult<Map<String, dynamic>>> getHistory(String id) =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentHistory(id),
        localCollection: 'students',
        localFilters: {'id': id},
      );

  // ─── Statistiques & Orion ────────────────────────────────────────────────

  /// Récupère les statistiques des élèves pour une année et niveau.
  ///
  /// Miroir de `studentsService.getStatistics(academicYearId, schoolLevelId)`
  /// du web app.
  Future<ApiResult<Map<String, dynamic>>> getStatistics(
    String academicYearId,
    String schoolLevelId,
  ) =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentStatistics(academicYearId, schoolLevelId),
      );

  /// Récupère les KPIs Orion pour une année académique.
  ///
  /// Miroir de `studentsService.getOrionKpis(academicYearId)` du web app.
  Future<ApiResult<Map<String, dynamic>>> getOrionKpis(
    String academicYearId,
  ) =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentOrionKpis(academicYearId),
      );

  /// Récupère les alertes Orion pour une année académique.
  ///
  /// Miroir de `studentsService.getOrionAlerts(academicYearId)` du web app.
  Future<ApiResult<List<Map<String, dynamic>>>> getOrionAlerts(
    String academicYearId,
  ) =>
      apiGetWithFallback<List<Map<String, dynamic>>>(
        ApiConfig.studentOrionAlerts(academicYearId),
        localCollection: 'students',
        localFilters: {'type': 'orionAlert'},
      );

  // ─── Photo & Gardiens ────────────────────────────────────────────────────

  /// Upload la photo d'un élève.
  ///
  /// Miroir de `studentsService.uploadPhoto(id, file)` du web app.
  Future<ApiResult<Map<String, dynamic>>> uploadPhoto(
    String id,
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentUploadPhoto(id),
        data,
      );

  /// Ajoute des gardiens (parents) à un élève.
  ///
  /// Miroir de `studentsService.addGuardians(id, data)` du web app.
  Future<ApiResult<Map<String, dynamic>>> addGuardians(
    String id,
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentGuardians(id),
        data,
      );

  // ─── Identifiants & Cartes ───────────────────────────────────────────────

  /// Récupère les statistiques des cartes d'identité.
  ///
  /// Miroir de `studentsService.getIdCardStats()` du web app.
  Future<ApiResult<Map<String, dynamic>>> getIdCardStats() =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentIdCardStats,
      );

  /// Génère des cartes d'identité en masse.
  ///
  /// Miroir de `studentsService.generateBulkIdCards(data)` du web app.
  Future<ApiResult<Map<String, dynamic>>> generateBulkIdCards(
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentGenerateBulkIdCards,
        data,
      );

  /// Récupère les statistiques des matricules.
  ///
  /// Miroir de `studentsService.getMatriculeStats()` du web app.
  Future<ApiResult<Map<String, dynamic>>> getMatriculeStats() =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentIdentifiersStats,
      );

  /// Génère des matricules en masse.
  ///
  /// Miroir de `studentsService.generateBulkMatricules(data)` du web app.
  Future<ApiResult<Map<String, dynamic>>> generateBulkMatricules(
    Map<String, dynamic> data,
  ) =>
      apiPostWithOfflineFallback<Map<String, dynamic>>(
        ApiConfig.studentGenerateBulkMatricules,
        data,
      );

  /// Recherche un élève par matricule.
  ///
  /// Miroir de `studentsService.searchByMatricule(matricule)` du web app.
  Future<ApiResult<Map<String, dynamic>>> searchByMatricule(
    String matricule,
  ) =>
      apiGetWithFallback<Map<String, dynamic>>(
        ApiConfig.studentSearchByMatricule(matricule),
        localCollection: 'students',
        localFilters: {'matricule': matricule},
      );
}

File: apps/web-app/src/services/students.service.ts
/**
 * ============================================================================
 * STUDENTS SERVICE - API Client pour le module Scolarité (Élèves)
 * ============================================================================
 */

import { apiFetch } from "@/lib/api/client";
import { createEntityOffline, updateEntityOffline } from "@/lib/offline/offline-business.service";
import { networkDetectionService } from "@/lib/offline/network-detection.service";
import { LocalSearchService } from "@/lib/offline/local-search.service";

const BASE_URL = "/students";

function getTenantId(): string {
  if (typeof document === "undefined") return "";
  const match = document.cookie.match(/(?:(?:^|.*;\s*)x-tenant-id\s*\=\s*([^;]*).*$)|^.*$/);
  return match ? decodeURIComponent(match[1]) : "";
}

class StudentsService {
  /**
   * Récupère la liste des étudiants
   */
  async getAll(params?: Record<string, string>): Promise<any> {
    const qs = params ? new URLSearchParams(params).toString() : "";
    
    if (!networkDetectionService.isConnected()) {
      return LocalSearchService.search("students", { tenantId: getTenantId() });
    }

    try {
      return await apiFetch(`${BASE_URL}${qs ? `?${qs}` : ""}`);
    } catch (error) {
      return LocalSearchService.search("students", { tenantId: getTenantId() });
    }
  }

  /**
   * Admissions — avec fallback offline
   */
  async getAdmissions(params?: Record<string, string>): Promise<any> {
    const qs = params ? new URLSearchParams(params).toString() : "";

    if (!networkDetectionService.isConnected()) {
      return LocalSearchService.search("students", { tenantId: getTenantId(), filters: { type: 'admission' } });
    }

    // ⚠️ NE PAS catcher silencieusement — si l'API retourne 500 (migrations
    // non appliquées, FK violation, etc.), l'utilisateur doit voir l'erreur
    // au lieu d'avoir une liste vide sans explication.
    return apiFetch(`${BASE_URL}/admissions${qs ? `?${qs}` : ""}`);
  }

  async getAdmissionById(id: string): Promise<any> {
    if (!networkDetectionService.isConnected()) {
      const results = await LocalSearchService.search("students", { tenantId: getTenantId() });
      return results.find((s: any) => s.id === id) || null;
    }
    try {
      return await apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}`);
    } catch (error) {
      const results = await LocalSearchService.search("students", { tenantId: getTenantId() });
      return results.find((s: any) => s.id === id) || null;
    }
  }

  async createAdmission(data: any): Promise<any> {
    // ⚠️ IMPORTANT : les admissions DOIVENT être sauvegardées dans la vraie DB
    // (pas en mode offline). Avant, le code vérifiait `if (tenantId)` et
    // redirigeait vers createEntityOffline — mais tenantId est TOUJOURS défini
    // en production → l'admission n'était jamais envoyée au backend.
    // Le toast disait "succès" mais la DB était vide.
    return apiFetch(`${BASE_URL}/admissions`, {
      method: "POST",
      body: data,
    });
  }

  async updateAdmission(id: string, data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: data,
    });
  }

  async submitAdmission(id: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/submit`, {
      method: "POST",
    });
  }

  async decideAdmission(
    id: string,
    data: { decision: "ACCEPTED" | "REJECTED"; comment: string },
  ): Promise<any> {
    // ⚠️ Utiliser fetch direct au lieu de apiFetch pour éviter les problèmes
    // de sérialisation axios. Le backend attend { decision, comment } en JSON.
    const apiUrl = process.env.NEXT_PUBLIC_API_URL || (typeof window !== 'undefined' && window.location.host.includes('academiahelm.com') ? 'https://api.academiahelm.com/api' : '');
    const url = `${apiUrl}/students/admissions/${encodeURIComponent(id)}/decide`;
    const token = typeof window !== 'undefined' ? localStorage.getItem('access_token') : null;
    const res = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      credentials: 'include',
      body: JSON.stringify({ decision: data.decision, comment: data.comment }),
    });
    if (!res.ok) {
      const err = await res.json().catch(() => ({}));
      throw new Error(err.message || err.error || `HTTP ${res.status}`);
    }
    return res.json();
  }

  async convertAdmission(id: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/admissions/${encodeURIComponent(id)}/convert`,
      {
        method: "POST",
      },
    );
  }

  async deleteAdmission(id: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}`, {
      method: "DELETE",
    });
  }

  async acceptAdmission(id: string, comment?: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/accept`, {
      method: "POST",
      body: { comment },
    });
  }

  async rejectAdmission(id: string, comment?: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/reject`, {
      method: "POST",
      body: { comment },
    });
  }

  async waitlistAdmission(id: string, comment?: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/waitlist`, {
      method: "POST",
      body: { comment },
    });
  }

  async cancelAdmission(id: string, comment?: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/cancel`, {
      method: "POST",
      body: { comment },
    });
  }

  async requestDocuments(id: string, comment?: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(id)}/request-documents`, {
      method: "POST",
      body: { comment },
    });
  }

  // ─── Admission Documents ──────────────────────────────────────────────────
  async getAdmissionDocuments(admissionId: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(admissionId)}/documents`);
  }

  async createAdmissionDocument(admissionId: string, data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(admissionId)}/documents`, {
      method: "POST",
      body: data,
    });
  }

  async validateAdmissionDocument(documentId: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/documents/${encodeURIComponent(documentId)}/validate`, {
      method: "POST",
    });
  }

  async rejectAdmissionDocument(documentId: string, comment?: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/documents/${encodeURIComponent(documentId)}/reject`, {
      method: "POST",
      body: { comment },
    });
  }

  async deleteAdmissionDocument(documentId: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/documents/${encodeURIComponent(documentId)}`, {
      method: "DELETE",
    });
  }

  // ─── Admission Interviews ─────────────────────────────────────────────────
  async getAdmissionInterviews(admissionId: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(admissionId)}/interviews`);
  }

  async createAdmissionInterview(admissionId: string, data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/${encodeURIComponent(admissionId)}/interviews`, {
      method: "POST",
      body: data,
    });
  }

  async completeAdmissionInterview(interviewId: string, data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/interviews/${encodeURIComponent(interviewId)}/complete`, {
      method: "POST",
      body: data,
    });
  }

  async deleteAdmissionInterview(interviewId: string): Promise<any> {
    return apiFetch(`${BASE_URL}/admissions/interviews/${encodeURIComponent(interviewId)}`, {
      method: "DELETE",
    });
  }

  /**
   * Crée un nouvel étudiant
   */
  async create(data: any): Promise<any> {
    // Mode Offline-First complet pour les opérations CRUD basiques
    const tenantId = getTenantId();
    if (tenantId) {
      return createEntityOffline(tenantId, "STUDENT", data);
    }
    
    // Fallback si tenantId non trouvé
    return apiFetch(BASE_URL, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Met à jour un étudiant
   */
  async update(id: string, data: any): Promise<any> {
    const tenantId = getTenantId();
    if (tenantId) {
      return updateEntityOffline(tenantId, "STUDENT", id, data);
    }
    
    // Fallback si tenantId non trouvé
    return apiFetch(`${BASE_URL}/${encodeURIComponent(id)}`, {
      method: "PATCH",
      body: data,
    });
  }

  /**
   * Récupère les inscriptions — avec fallback offline
   */
  async getEnrollments(params?: Record<string, string>): Promise<any> {
    const qs = params ? new URLSearchParams(params).toString() : "";
    
    if (!networkDetectionService.isConnected()) {
      return LocalSearchService.search("students", { tenantId: getTenantId(), filters: { enrollmentStatus: 'ENROLLED' } });
    }
    
    try {
      return await apiFetch(`${BASE_URL}/enrollments${qs ? `?${qs}` : ""}`);
    } catch (error) {
      return LocalSearchService.search("students", { tenantId: getTenantId(), filters: { enrollmentStatus: 'ENROLLED' } });
    }
  }

  /**
   * Crée une inscription / admission — offline-first
   */
  async createEnrollment(data: any): Promise<any> {
    const tenantId = getTenantId();
    if (tenantId) {
      return createEntityOffline(tenantId, "STUDENT", { ...data, type: 'enrollment' });
    }
    return apiFetch(`${BASE_URL}/enrollments`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Pré-inscription
   */
  async preRegister(data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/pre-register`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Ajouter des tuteurs
   */
  async addGuardians(
    studentId: string,
    data: { guardians: any[] },
  ): Promise<any> {
    return apiFetch(`${BASE_URL}/${encodeURIComponent(studentId)}/guardians`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Finaliser l'inscription (enroll)
   */
  async enrollStudent(studentId: string, data: any): Promise<any> {
    return apiFetch(`${BASE_URL}/${encodeURIComponent(studentId)}/enroll`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Statistiques des cartes d'identité
   */
  async getIdCardStats(academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/id-cards/stats?academicYearId=${academicYearId}`,
    );
  }

  /**
   * Génération par lot de cartes d'identité
   */
  async generateBulkIdCards(data: {
    academicYearId: string;
    schoolLevelId: string;
  }): Promise<any> {
    return apiFetch(`${BASE_URL}/id-cards/generate-bulk`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Statistiques des matricules
   */
  async getMatriculeStats(): Promise<any> {
    return apiFetch(`${BASE_URL}/identifiers/stats`);
  }

  /**
   * Génération par lot de matricules
   */
  async generateBulkMatricules(data: {
    academicYearId: string;
    schoolLevelId: string;
    status?: string;
  }): Promise<any> {
    return apiFetch(`${BASE_URL}/identifiers/generate-bulk?countryCode=BJ`, {
      method: "POST",
      body: data,
    });
  }

  /**
   * Générer un matricule individuel
   */
  async generateMatricule(studentId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/identifiers/${encodeURIComponent(studentId)}/generate?countryCode=BJ`,
      {
        method: "POST",
      },
    );
  }

  /**
   * Rechercher par matricule
   */
  async searchByMatricule(matricule: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/identifiers/search/${encodeURIComponent(matricule)}`,
    );
  }

  /**
   * Statistiques Générales
   */
  async getStatistics(
    academicYearId: string,
    schoolLevelId: string,
  ): Promise<any> {
    const params = new URLSearchParams({ academicYearId, schoolLevelId });
    return apiFetch(`${BASE_URL}/statistics?${params}`);
  }

  /**
   * KPIs ORION
   */
  async getOrionKpis(academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/orion/kpis?academicYearId=${encodeURIComponent(academicYearId)}`,
    );
  }

  /**
   * Alertes ORION
   */
  async getOrionAlerts(academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/orion/alerts?academicYearId=${encodeURIComponent(academicYearId)}`,
    );
  }

  /**
   * Révoquer une carte d'identité
   */
  async revokeIdCard(id: string, reason: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/id-cards/card/${encodeURIComponent(id)}/revoke`,
      {
        method: "PUT",
        body: { reason },
      },
    );
  }

  /**
   * Dossier et historique
   */
  async getDossier(id: string, academicYearId?: string): Promise<any> {
    const url = `/api/students/${id}/dossier${academicYearId ? `?academicYearId=${academicYearId}` : ""}`;
    return apiFetch(url);
  }

  async getHistory(id: string): Promise<any> {
    return apiFetch(`${BASE_URL}/${encodeURIComponent(id)}/history`);
  }

  async getVerificationQR(id: string, academicYearId: string): Promise<any> {
    return apiFetch(
      `${BASE_URL}/${encodeURIComponent(id)}/verification-qr?academicYearId=${encodeURIComponent(academicYearId)}`,
    );
  }

  async regenerateVerificationQR(
    id: string,
    academicYearId: string,
  ): Promise<any> {
    return apiFetch(
      `${BASE_URL}/${encodeURIComponent(id)}/verification-token/regenerate`,
      {
        method: "POST",
        body: { academicYearId },
      },
    );
  }

  /**
   * Télécharge le dossier académique PDF
   */
  async downloadAcademicDossier(
    id: string,
    academicYearId: string,
  ): Promise<void> {
    const res = await fetch(
      `/api/students/${encodeURIComponent(id)}/academic-dossier?academicYearId=${encodeURIComponent(academicYearId)}`,
      {
        headers: {
          // Need to add auth headers manually if not using apiClient for blob
          "x-tenant-id": localStorage.getItem("tenantId") || "",
        },
      },
    );
    if (!res.ok) throw new Error("Erreur lors du téléchargement");
    const blob = await res.blob();
    const url = window.URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `DOSSIER_${id}.pdf`;
    document.body.appendChild(a);
    a.click();
    window.URL.revokeObjectURL(url);
    a.remove();
  }

  /**
   * Export Educmaster Excel
   */
  async exportEducmasterExcel(
    academicYearId: string,
    schoolLevelId: string,
  ): Promise<void> {
    const res = await fetch(
      `/api/students/export/educmaster-excel?academicYearId=${encodeURIComponent(academicYearId)}&schoolLevelId=${encodeURIComponent(schoolLevelId)}`,
      {
        headers: {
          "x-tenant-id": localStorage.getItem("tenantId") || "",
        },
      },
    );
    if (!res.ok) {
      const err = await res
        .json()
        .catch(() => ({ message: "Erreur lors de l'export" }));
      throw new Error(err.message || "Erreur lors de l'export");
    }
    const blob = await res.blob();
    const url = window.URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `EDUCMASTER_EXPORT_${schoolLevelId}_${academicYearId}.xlsx`;
    document.body.appendChild(a);
    a.click();
    window.URL.revokeObjectURL(url);
    a.remove();
  }
}

export const studentsService = new StudentsService();
